File: QuEST/CTools/newMemorymeasure.c
/** @file 
 * Reads your C's mac process' memory usage
 */
 
 
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>

#include "newMemorymeasure.h"


/*
 * Measures the current (and peak) resident and virtual memory
 * usage of your linux C process, in bytes, accurate to nearest kB.
 * Returns a 0 if memory info access was successful, else prints
 * an error message and returns 1
 */

// in size of Bytes
unsigned long sizeStringToInteger(char *string) {
    char unit = string[strlen(string)-1];
    char value[10];
    strncpy(value, string, strlen(string)-1);
    switch(unit) {
        case 'K': return (unsigned long) atof(value)*1024;
        case 'M': return (unsigned long) atof(value)*1024*1024;
        case 'G': return (unsigned long) atof(value)*1024*1024*1024;
        default: return (unsigned long) atof(value);
    }
}

int getMemory(
	unsigned long *currRealMem, unsigned long *peakRealMem,
	unsigned long *currVirtMem, unsigned long *peakVirtMem) {
	
	// stores each word in status file
	char buffer[1024] = "";
	
	// linux file contains this-process info
    char cmd[100];
    sprintf(cmd, "vmmap %d -summary > memoryFile", getpid());
    int status = system(cmd);

	FILE* file = NULL;
	file = fopen("memoryFile", "r");
	if (file == NULL) {
		printf("Call to getMemory FAILED; "
			   "linux file proc/self/status not found!\n");
		return 1;
	}

    char virtSize[10];
    char realSize[10];
	// read the entire file, recording mems in kB
    while(fscanf(file, "%1023s", buffer) == 1) {
        if (strcmp(buffer, "TOTAL") == 0) {
            fscanf(file, "%s", virtSize);
            fscanf(file, "%s", realSize);
            *currVirtMem = sizeStringToInteger(virtSize);
            *currRealMem = sizeStringToInteger(realSize);
            *peakVirtMem = sizeStringToInteger(virtSize);
            *peakRealMem = sizeStringToInteger(realSize);
        }
    }
    fclose(file);

    system("rm memoryFile");
	return 0;
}


// int main(int narg, char **argv) {
//     unsigned long a = 0;
//     unsigned long b = 0;
//     unsigned long c = 0;
//     unsigned long d = 0;
//     getMemory(&a, &b, &c, &d);
//     return 0;
// }

File: NewBenchmark/Python/new_memory_measure.py

_FIELDS = ['VmRSS', 'VmHWM', 'VmSize', 'VmPeak']

import os

def sizeStirngToInteger(sizeString):
    unit = sizeString[-1]
    value = sizeString[:-1]
    if unit == "K":
        return int(float(value)*1024)
    elif unit == "M":
        return int(float(value)*1024*1024)
    elif unit == "G":
        return int(float(value)*1024*1024*1024)
    else:
        return int(float(value))


def get_memory():
    '''
    returns the current and peak, real and virtual memories
    used by the calling linux python process, in Bytes
    '''

    pid = os.getpid()
    cmd = "vmmap %d -summary > memoryFile"%pid
    os.system(cmd)

    try:
        with open("memoryFile", "r") as file:
            for line in file:
                linesplit = line.split()
                if len(linesplit) > 0 and linesplit[0] == "TOTAL":
                    virtSize = linesplit[1]
                    realSize = linesplit[2]
    except:
        print("Call to getMemory FAILDED; mac file memoryFile not found!\n")
        return 1

    mem = {}
    mem["VmRSS"] = sizeStirngToInteger(realSize)
    mem["VmSize"] = sizeStirngToInteger(virtSize)
    mem["VmHWM"] = sizeStirngToInteger(realSize)
    mem["VmPeak"] = sizeStirngToInteger(virtSize)
    return mem


if __name__ == '__main__':

    # a simple test
    print(get_memory())
    mylist = [1.5]*2**10  # 30
    print(get_memory())
    del mylist
    print(get_memory())
